src/validation.cpp

```diff
@@ -4930,8 +4931,22 @@ void ChainstateManager::CheckBlockIndex()
             assert(((pindex->nStatus & BLOCK_VALID_MASK) >= BLOCK_VALID_TRANSACTIONS) == (pindex->nTx > 0)); // This is pruning-independent.
         }
         // All parents having had data (at some point) is equivalent to all parents being VALID_TRANSACTIONS, which is equivalent to HaveNumChainTxs().
+        // However, an exception to this is possible for the snapshot base
+        // block. In rare cases, the snapshot base block, which has a non-zero
+        // nChainTx value initialized from snapshot metadata, may temporarily
+        // have its nChainTx value set to 0 by ReceivedBlockTransations if it is
+        // downloaded out of order, before one of its parent blocks. This is
+        // confusing behavior, but not a bug as long as the check below is
+        // written to account for it. More work should be done later to clean up
+        // the behavior and simplify the check, see
+        // https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505114232
         assert((pindexFirstNeverProcessed == nullptr) == pindex->HaveNumChainTxs());
-        assert((pindexFirstNotTransactionsValid == nullptr) == pindex->HaveNumChainTxs());
+        if (pindex->HaveNumChainTxs()) {
+            assert(pindexFirstNotTransactionsValid == nullptr);
+        } else {
+            const bool unprocessed_snapshot_block{pindex == snap_base && pindex->nTx == 0};
+            assert(pindexFirstNotTransactionsValid != nullptr || unprocessed_snapshot_block);
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1509552384

> I think it must be `pindex->nTx != 0` in the line above - since the situation is that we have received the snapshot block transactions (so `nTx!=0`), but can't connect the block yet (so `nChainTx==0`). Also, maybe `unconnected_snapshot_block` is a better name?

Oh, I forgot that nTx would still be set in that case. I was trying to make the check stricter so snap_base would have to have nChainTx set after it was fully connected, and it wouldn't be able to escape this check after that happened. But I guess there is no way to determine that and the best that can be done is to check that it nChainTx is nonzero after it is validated. Can make that update

Reply:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1509552384

Could reproduce this failure with the test change suggested in https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1974096225. Changing this to `const bool unprocessed_snapshot_block{pindex == snap_base && (pindex->nStatus & BLOCK_VALID_MASK) < BLOCK_VALID_SCRIPTS}` did fix this but then there was just another assert failure in the `assert((pindex->nChainTx != 0) == (pindex == snap_base));` line below, so opted to revert this and make a fix in `ReceivedBlockTransactions` instead